squidpy/pl/_spatial_utils.py

```diff
@@ -0,0 +1,897 @@
+    spatial_key: str = Key.obsm.spatial,
+    library_id: Sequence[str] | None = None,
+    library_key: str | None = None,
+    img: _SeqArray | bool = None,
+    img_res_key: str | None = Key.uns.image_res_key,
+    img_channel: int | None = None,
+    seg: _SeqArray | bool = None,
+    seg_key: str | None = None,
+    cell_id_key: str | None = None,
+    scale_factor: _SeqFloat = None,
+    size: _SeqFloat = None,
+    size_key: str | None = Key.uns.size_key,
+    img_cmap: Colormap | str | None = None,
+) -> SpatialParams:
+
+    library_id = _get_library_id(
+        adata=adata, shape=shape, spatial_key=spatial_key, library_id=library_id, library_key=library_key
+    )
+
+    scale_factor, size = _get_scalefactor_size(
+        adata=adata,
+        spatial_key=spatial_key,
+        library_id=library_id,
+        img_res_key=img_res_key,
+        scale_factor=scale_factor,
+        size=size,
+        size_key=size_key,
+    )
+
+    if img:
+        _img = _get_image(
+            adata=adata,
+            spatial_key=spatial_key,
+            library_id=library_id,
+            img=img,
+            img_res_key=img_res_key,
+            img_channel=img_channel,
+            img_cmap=img_cmap,
+        )
+    else:
+        _img = tuple(None for _ in library_id)
+
+    if seg:
+        _seg, _cell_vec = _get_segment(
+            adata=adata,
+            library_id=library_id,
+            cell_id_key=cell_id_key,
+            library_key=library_key,
+            seg=seg,
+            seg_key=seg_key,
+        )
+    else:
+        _seg = tuple(None for _ in library_id)
+        _cell_vec = tuple(None for _ in library_id)
+
+    return SpatialParams(library_id, scale_factor, size, _img, _seg, _cell_vec)
+
+
+def _set_coords_crops(
+    adata: AnnData,
+    spatial_params: SpatialParams,
+    spatial_key: str,
+    library_key: str | None = None,
+    crop_coord: Sequence[_CoordTuple] | _CoordTuple | None = None,
+) -> Tuple[Sequence[NDArrayA], list[Tuple[float, ...]] | Tuple[None, ...]]:
+
+    # set crops
+    if crop_coord is None:
+        crops: Union[list[Tuple[float, ...]], Tuple[None, ...]] = tuple(None for _ in spatial_params.library_id)
+    else:
+        crop_coord = _get_list(crop_coord, tuple, len(spatial_params.library_id))
+        crops = [CropCoords(*cr) * sf for cr, sf in zip(crop_coord, spatial_params.scale_factor)]
+
+    coords = adata.obsm[spatial_key]
+    if library_key is None:
+        return [coords * sf for sf in spatial_params.scale_factor], crops
+    else:
+        return [
+            coords[adata.obs[library_key] == lib, :] * sf  # TODO: check that library_key is asserted upstream
+            for lib, sf in zip(spatial_params.library_id, spatial_params.scale_factor)
+        ], crops
+
+
+def _subs(adata: AnnData, library_key: str | None = None, library_id: str | None = None) -> AnnData:
+    try:
+        return adata[adata.obs[library_key] == library_id].copy()
```

Comment:
Is the copy necessary? I guess it is, just asking, can be problematic making a lots of hidden copies just for plotting.

Reply:
yeah it is, do you think it'd clutter the memory session? a solution for that?

Reply:
I've look at the place where you call `_subs`, some suggestions:
- you call it twice (once when calling `_set_color_source_vec`, other time in `_plot_edges`), should be called only once
- as for copying, the problem mostly lies with `AnnData` when you tried to write to a view (it creates a copy for you)
  looking at `_set_color_source_vec`, you don't write anything to `adata`, so it should be fine
  as for `_plot_edges`, it's the same story
  
so I think the copy can be removed (unless I'm missing something).

Reply:
good point, removed